File: new_practice_python/103.binary-tree-zigzag-level-order-traversal.py
#
# @lc app=leetcode id=103 lang=python3
#
# [103] Binary Tree Zigzag Level Order Traversal
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    # def zigzagLevelOrder(self, root: TreeNode) -> List[List[int]]:
    def zigzagLevelOrder(self, root):  

        ans = []
        level = [root]
        Zigzag = True

        if root is None:
            return []

        while level:
            # add new level
            ans_temp = []
            for ele in level:
                ans_temp.append(ele.val)


            if Zigzag:
                ans.append(ans_temp)
            else:
                ans.append(ans_temp[::-1])

            Zigzag = not Zigzag

            # find next Level
            next_level = []
            for ele in level:
                if ele.left is not None:
                    next_level.append(ele.left)
                if ele.right is not None:   
                    next_level.append(ele.right)

            level = next_level

        return ans


        
# @lc code=end

File: new_practice_cpp/103.binary-tree-zigzag-level-order-traversal.cpp
/*
 * @lc app=leetcode id=103 lang=cpp
 *
 * [103] Binary Tree Zigzag Level Order Traversal
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {

        if(!root) return {};

        vector<vector<int>> ans;
        vector<TreeNode*> level = {root};
        bool zigzag = true;

        while(!level.empty()){

            vector<int> ans_level = {};
            if(zigzag){
                for(int i = 0;i< level.size();i++)
                    ans_level.push_back(level[i]->val);
            }
            else{
                for(int i = level.size()-1;i>=0;i--)
                    ans_level.push_back(level[i]->val);
            }
            zigzag = !zigzag;
            ans.push_back(ans_level);

            vector<TreeNode*> next_level = {};
            for(int i = 0;i< level.size();i++){
                if(level[i]->left) // if exists
                    next_level.push_back(level[i]->left); 
                if(level[i]->right) // if exists
                    next_level.push_back(level[i]->right); 
            }
            level = next_level;
        }
        return ans;
    }
};
// @lc code=end

